### ISSUE TITLE
Traccar API on 3.3

### ISSUE THREAD

**Issue by mofesola** [tananaev/traccar#1616]
Hi,

Great work on this new release. I can see a lot of  bugs relating to my NoranProtocol device have been fixed. I however need to understand how to use the new API. I can't seem to call from my Rest Clients anymore, Also I cannot seem to login. Any help will be appreciated.

**Comment by tananaev** [comment]
The API has changed. I am working on the documentation for it.

**Comment by mofesola** [comment]
Hi,

Do we have any updates on this now, as we really need to reinstate our API connection.


Any rough ideas will be of help.

**Comment by tananaev** [comment]
I have create a swagger file that you can use as a documentation for the new API. Here you can find more details and reference to the file:

https://www.traccar.org/traccar-api/

**Comment by mofesola** [comment]
Thanks. I however get permission denied when I use /api/positions

/api/session shows a session has been created and I should have access.

**Comment by tananaev** [comment]
Are you using the cookie returned with session request?

**Comment by mofesola** [comment]
No, I am not. I tried using it, and I still get the same error

{
"details": "Device access denied - SecurityException (PermissionsManager:89 < PositionResource:41 < ...)"
"message": "Device access denied"
}

**Comment by tananaev** [comment]
Now it looks like the user doesn't have access to the device.

**Comment by mofesola** [comment]
But I do, from traccar web client, I can see the device

**Comment by tananaev** [comment]
Can I see the request that you are sending?

**Comment by mofesola** [comment]
GET http://<SERVER>/api/positions

deviceId: 10
JSESSIONID: 1uzpbi1j736ry1c2jrtafbqygq

**Comment by tananaev** [comment]
Try to get devices and see if user has access to this particular device. Also, you are missing other parameters (`from` and `to` timestamps).

**Comment by mofesola** [comment]
Yes, when I ran /api/devices, I got: a list of devices including this:

3:  {
"id": 10
"name": "Fuel Level Pico"
"uniqueId": "NR09G13609"
"status": "online"
"lastUpdate": "2016-01-12T08:10Z"
"positionId": 84396
"dataId": 0
}

**Comment by tananaev** [comment]
It doesn't looks like you are using version 3.3.

**Comment by mofesola** [comment]
I have 3.3 installed, checked the db changelog, The latest file there is 3.3

**Comment by tananaev** [comment]
Version 3.3 doesn't have `dataId` parameter for devices.

**Comment by mofesola** [comment]
How can I confirm the version of traccar installed, apart from what I did there?

I'm certain I did a clean install of 3.3

**Comment by tananaev** [comment]
OK, I was wrong. The `dataId` field was there, but it wasn't used.

I would recommend to look at what requests default web interface sends and compare it to your ones.

**Comment by AYastrebov** [comment]
Can I connect with websocket? How should I do it?

**Comment by tananaev** [comment]
You need to create a user session first using regular AP. After that you can connection to the WebSocket (`/api/socket`).

**Comment by mofesola** [comment]
Here's my request monitored:

**Comment by tananaev** [comment]
Have you trie to compare it with Traccar web interface requests? There must something missing.

**Comment by mofesola** [comment]
How can I get what the web interface is sending out?

**Comment by tananaev** [comment]
You can use browser developer tools.

**Comment by zzheads** [comment]
How I can get ALL positions? Using GET http://myserver/api/positions without any parameters (as said in API docs) doesn't work - I get positions, but not all. Found it when try find position from event reference (by positionId), so can't find that position in positions previously fetched from server (there must be ALL), but found it on server by id.

**Comment by tananaev** [comment]
@zzheads, are you asking of ALL position from the database? Obviously there is no API for that. You can only get all latest positions for your devices.

**Comment by zzheads** [comment]
What means latest? For what period of time? Position I was unable to find was recent.

**Comment by tananaev** [comment]
Latest means latest for each device. No period of time.

**Comment by zzheads** [comment]
1 position for each device?

**Comment by tananaev** [comment]
Yes.

**Comment by zzheads** [comment]
Thanks

**Comment by md-farhan-memon** [comment]
Hi @tananaev 
I was trying to fetch latest positions of devices, as per the documentation I used `/positions` without any parameters but I get `400` response code with error `NullPointerException (... < DateUtil:64 < PositionResource:80 < ...)`.
I even tried adding `id` as `positionId` which I get from `/devices` api but same error. Any change in documentation or APIs?
I am on `3.14-SNAPSHOT`

**Comment by md-farhan-memon** [comment]
Nevermind, found the issue.
One needs to add `Accept` header else it throws the above error. I believe the error message should be more relevant in this case, shouldn't it?
Anyway, credits to this [post](https://www.traccar.org/forums/topic/error-400-api-positions/) and your comment there.
Hope it will help others who stumble upon here while googling.. :)
